src/Fabulous/Primitives.fs

```diff
@@ -1,57 +1,177 @@
+        fun msg ->
+            let mutable parentOpt = node.Parent
+            let mutable mapMsg =
+                match node.MapMsg with
+                | ValueNone -> id
+                | ValueSome fn -> fn
+
+            while parentOpt.IsSome do
+                let parent = parentOpt.Value :?> IViewNodeWithEvents
+                parentOpt <- parent.Parent
+
+                mapMsg <-
+                    match parent.MapMsg with
+                    | ValueNone -> mapMsg
+                    | ValueSome fn -> mapMsg >> fn
+
+            let newMsg = mapMsg msg
+            //node.TreeContext.Dispatch(newMsg)
+            ()
+
+    interface IEventAttributeDefinition with
+        member x.Key = x.Key
+        
+        member x.CompareBoxed a b =
+            x.Compare (unbox<'inputType> a) (unbox<'inputType> b)
+        
+        member x.RemoveHandler node =
+            match node.TryGetHandler(x) with
+            | ValueNone -> ()
+            | ValueSome handler ->
+                let event = x.GetEvent(unbox node.Target)
+                event.RemoveHandler(handler)
+            
+        member x.AddHandler newValueOpt node =
+            match newValueOpt with
+            | ValueNone ->
+                node.SetHandler(x, ValueNone)
+            | ValueSome v ->
+                let msgFn = x.Convert(unbox v)
+                let dispatchFn = msgFn >> x.DispatchMsg node
+                let handler = x.ConvertValue(dispatchFn)
+                let event = x.GetEvent(unbox node.Target)
+                node.SetHandler(x, ValueSome handler)
+                event.AddHandler(handler)
+
 [<Struct>]
 type ScalarAttribute =
-    { Key: AttributeKey
-#if DEBUG
-      DebugName: string
-#endif
+    { Definition: IScalarAttributeDefinition
+      Value: obj }
+
+[<Struct>]
+type EventAttribute =
+    { Definition: IEventAttributeDefinition
       Value: obj }
 
+type [<ReferenceEquality>] WidgetAttributeDefinition =
+    { Key: string
+      GetChildNode: IViewNode -> IViewNode
+      CreateNode: Widget -> IViewNode -> unit
+      RemoveNode: IViewNode -> unit }
+
+and [<ReferenceEquality>] WidgetCollectionAttributeDefinition =
+    { Key: string
+      GetItemNode: IViewNode -> int -> IViewNode
+      Insert: IViewNode -> int -> obj -> unit
+      Replace: IViewNode -> int -> obj -> unit
+      Remove: IViewNode -> int -> unit
+      UpdateNode: ArraySlice<Widget> voption -> IViewNode -> unit }
 
 and [<Struct>] WidgetAttribute =
-    { Key: AttributeKey
-#if DEBUG
-      DebugName: string
-#endif
+    { Definition: WidgetAttributeDefinition
       Value: Widget }
-
+    
 and [<Struct>] WidgetCollectionAttribute =
-    { Key: AttributeKey
-#if DEBUG
-      DebugName: string
-#endif
+    { Definition: WidgetCollectionAttributeDefinition
       Value: ArraySlice<Widget> }
 
-/// Represents a virtual UI element such as a Label, a Button, etc.
+and [<Struct>] WidgetData =
+    { ScalarAttributes: ScalarAttribute[] voption
```

Comment:
also it is possible that we might want to optimize `voption` in here in the future. Maybe with direct null checks.
voption  adds 4 bytes overhead :( Or maybe build ourselves `TinyOption` that is based on bool with custom pattern matching